### DISCUSSION TITLE
How to modify my url load function to wait not for "completed" but just "interactive" loading state?

### DISCUSSION THREAD

**Discussion by assopri** [cefsharp/CefSharp#3558] [Q&A]
I have following function to load webpage. It is waiting for "completed" state of document to return. However in my case "interactive" state (https://developer.mozilla.org/en-US/docs/Web/API/Document/readyState) is enough. How to modify my url load function to wait not for "completed" but just "interactive" loading state?

`public virtual void LoadUrlCore(string url)
        {
            if (String.IsNullOrEmpty(url)) return;

            var tcs = new TaskCompletionSource<bool>();
            EventHandler<LoadingStateChangedEventArgs> handler = null;
            handler = (sender, args) =>
            {
                if (!args.IsLoading)
                {
                    Browser.LoadingStateChanged -= handler;
                    tcs.TrySetResultAsync(true);
                }
            };
        
            Browser.LoadingStateChanged += handler;
            if (!string.IsNullOrEmpty(url))
            {
                Browser.Load(url);
            }
            

            if (!tcs.Task.Wait(TimeSpan.FromMilliseconds(LoadTimeoutMs)))
                throw new TimeoutException("Load timeout");
        }`

**Comment by amaitland** [comment]
The three events you can use are:

- http://cefsharp.github.io/api/90.6.x/html/E_CefSharp_IWebBrowser_FrameLoadEnd.htm
- http://cefsharp.github.io/api/90.6.x/html/E_CefSharp_IWebBrowser_FrameLoadStart.htm
- http://cefsharp.github.io/api/90.6.x/html/E_CefSharp_IWebBrowser_LoadingStateChanged.htm

If you need to execute javascript at a particular readyState then you'll need to use javascript to achieve that See https://github.com/cefsharp/CefSharp/wiki/General-Usage#when-can-i-start-executing-javascript

```c#
//Wait for the MainFrame to finish loading
browser.FrameLoadStart+= (sender, args) =>
{
  //MainFrame has started to load, too early to access the DOM, you can add event listeners for DOMContentLoaded etc.
  if(args.Frame.IsMain)
  {
    const string script = "document.addEventListener('DOMContentLoaded', function(){ alert('DomLoaded'); });";

    args.Frame.ExecuteJavaScriptAsync(script);
  }
};
```

You can use something like https://developer.mozilla.org/en-US/docs/Web/API/Document/readyState#readystatechange_as_event_listener_to_insert_or_modify_the_dom_before_domcontentloaded to pick the particular readyState you are after to execute your code.

If your just interested in the page load state then picking one of the three events above is the easiest option.

> if (!tcs.Task.Wait(TimeSpan.FromMilliseconds(LoadTimeoutMs)))
>            throw new TimeoutException("Load timeout");

The browser is designed to be `async`, please use `await` or `ContinueWith` rather than blocking the thread. Whilst this may work it is not supported.

**Reply by assopri** [reply]
Thank you for your reply. However I didn`t get how exactly can I stop further loading process in the LoadingStateChanged (when I detected the needed stage already occured). Smth similar to e.Cancel = true or smth like that?

**Reply by amaitland** [reply]
No such property exists.

http://cefsharp.github.io/api/90.6.x/html/Properties_T_CefSharp_LoadingStateChangedEventArgs.htm
http://cefsharp.github.io/api/90.6.x/html/Methods_T_CefSharp_IBrowser.htm

**Reply by assopri** [reply]
ok than I will try to use StopLoad when got stage, which is enough for my needs. Thank you!

**Comment by assopri** [comment]
However I still didn`t find solution on how to stop loading on "interactive" readystate. LoadingStateChanged just doesnt catch it. Let`s try from another side. My goal is to avoid loading sub-resources such as scripts, images, stylesheets and frames. So eventually I could get webpages to load faster. Can I somehow implement that?

![image](https://user-images.githubusercontent.com/73235464/119080524-e8996500-ba02-11eb-8ccb-c4dc6a7142ae.png)

**Reply by amaitland** [reply]
> My goal is to avoid loading sub-resources such as scripts, images, stylesheets and frames.

Yes, you can cancel loading requests quite easily see http://cefsharp.github.io/api/90.6.x/html/M_CefSharp_Handler_ResourceRequestHandler_OnBeforeResourceLoad.htm

Follows the same pattern as in the example at https://github.com/cefsharp/CefSharp/wiki/General-Usage#useragent

Just cancel the resource based on http://cefsharp.github.io/api/90.6.x/html/P_CefSharp_IRequest_ResourceType.htm

In future please start with what you are actually trying to achieve, save me a lot of time.

**Reply by assopri** [reply]
Thank you! However somewhy I cannot attach onbeforeresourceload correctly. I tried to do that like here: https://github.com/amaitland/CefSharp.MinimalExample/blob/demo/OnBeforeResourceLoad/CefSharp.MinimalExample.WinForms/BrowserForm.cs
However in debug I don not get into OnBeforeResourceLoad

```c#
public class CustomResourceRequestHandler : CefSharp.Handler.ResourceRequestHandler
    {
        protected override CefReturnValue OnBeforeResourceLoad(IWebBrowser chromiumWebBrowser, IBrowser browser, IFrame frame, IRequest request, IRequestCallback callback)
        {
            //my logic gonna live here

            return CefReturnValue.Continue;
        }
    }

    public class CustomRequestContextHandler : RequestHandler
    {

        protected override IResourceRequestHandler GetResourceRequestHandler(IWebBrowser chromiumWebBrowser, IBrowser browser, IFrame frame, IRequest request, bool isNavigation, bool isDownload, string requestInitiator, ref bool disableDefaultHandling)
        {
            return new CustomResourceRequestHandler();
        }
        

       
    }
```

**Reply by amaitland** [reply]
Did you assigned a new instance of your `RequestHandler` implementation to the `ChromiumWebBrowser.RequestHandler` property?

I'd suggest renaming `CustomRequestContextHandler` to `CustomRequestHandler` as that better suites it's purpose.

https://github.com/amaitland/CefSharp.MinimalExample/blob/demo/OnBeforeResourceLoad/CefSharp.MinimalExample.WinForms/BrowserForm.cs#L32

**Reply by assopri** [reply]
Yep, you were completely correct - I just missed new instance assignement. Thank you once more!
